File: scala/day16_part1_2019.scala
object Solution extends App {
  import scala.io.Source

  val source = Source.fromFile("input.txt")
  val input = source.getLines().mkString
  source.close()

  var digits = input.map(_.asDigit).toArray

  for (_ <- 0 until 100) {
    digits = applyFFT(digits)
  }

  println(digits.take(8).mkString)
}

def applyFFT(input: Array[Int]): Array[Int] = {
  val basePattern = Array(0, 1, 0, -1)
  val output = new Array[Int](input.length)

  for (i <- input.indices) {
    var sum = 0
    for (j <- input.indices) {
      val patternValue = basePattern(((j + 1) / (i + 1)) % basePattern.length)
      sum += input(j) * patternValue
    }
    output(i) = Math.abs(sum % 10)
  }

  output
}

File: perl/day16_part1_2019.pl

use strict;
use warnings;

open my $fh, '<', 'input.txt' or die $!;
my $input = <$fh>;
chomp $input;
close $fh;

my @digits = split //, $input;

for (1..100) {
    @digits = apply_fft(\@digits);
}

print join('', @digits[0..7]), "\n";

sub apply_fft {
    my ($input) = @_;
    my @base_pattern = (0, 1, 0, -1);
    my @output;

    for my $i (0 .. $#$input) {
        my $sum = 0;
        for my $j (0 .. $#$input) {
            my $pattern_value = $base_pattern[($j + 1) / ($i + 1) % @base_pattern];
            $sum += $input->[$j] * $pattern_value;
        }
        push @output, abs($sum) % 10;
    }
    return @output;
}

sub abs {
    return $_[0] < 0 ? -$_[0] : $_[0];
}
